runtime/lua/vim/lsp/snippet.lua

```diff
@@ -0,0 +1,406 @@
+    return factory()(input, pos)
+  end
+end
+
+P.token = function(token)
+  return function(input, pos)
+    local maybe_token = string.sub(input, pos, pos + #token - 1)
+    if token == maybe_token then
+      return {
+        parsed = true,
+        value = maybe_token,
+        pos = pos + #token,
+      }
+    end
+    return P.unmatch(pos)
+  end
+end
+
+P.pattern = function(p)
+  return function(input, pos)
+    local maybe_match = string.match(string.sub(input, pos), '^' .. p)
+    if maybe_match then
+      return {
+        parsed = true,
+        value = maybe_match,
+        pos = pos + #maybe_match,
+      }
+    end
+    return P.unmatch(pos)
+  end
+end
+
+P.many = function(parser)
+  return function(input, pos)
+    local values = {}
+    local new_pos = pos
+    while new_pos <= #input do
+      local result = parser(input, new_pos)
+      if not result.parsed then
+        break
+      end
+      table.insert(values, result.value)
+      new_pos = result.pos
+    end
+    if #values > 0 then
+      return {
+        parsed = true,
+        value = values,
+        pos = new_pos,
+      }
+    end
+    return P.unmatch(pos)
+  end
+end
+
+P.any = function(...)
+  local parsers = { ... }
+  return function(input, pos)
+    for _, parser in ipairs(parsers) do
+      local result = parser(input, pos)
+      if result.parsed then
+        return result
+      end
+    end
+    return P.unmatch(pos)
+  end
+end
+
+P.opt = function(parser)
+  return function(input, pos)
+    local result = parser(input, pos)
+    return {
+      parsed = true,
+      value = result.value,
+      pos = result.pos,
+    }
+  end
+end
+
+P.seq = function(...)
+  local parsers = { ... }
+  return function(input, pos)
+    local values = {}
+    local new_pos = pos
+    for _, parser in ipairs(parsers) do
+      local result = parser(input, new_pos)
+      if result.parsed then
+        table.insert(values, result.value)
+        new_pos = result.pos
+      else
+        return P.unmatch(pos)
+      end
+    end
+    return {
+      parsed = true,
+      value = values,
+      pos = new_pos,
+    }
+  end
+end
+
+--@see https://code.visualstudio.com/docs/editor/userdefinedsnippets#_grammar
+local S = {}
+S.dollar = P.token('$')
+S.open = P.token('{')
+S.close = P.token('}')
+S.colon = P.token(':')
+S.slash = P.token('/')
+S.comma = P.token(',')
+S.pipe = P.token('|')
+S.plus = P.token('+')
+S.minus = P.token('-')
+S.question = P.token('?')
+S.int = P.map(P.pattern('[0-9]+'), function(value)
+  return tonumber(value, 10)
+end)
+S.var = P.pattern('[%a_][%w_]+')
+S.text = function(targets, specials)
+  return P.map(P.take_until(targets, specials), function(value)
+    return {
+      type = 'text',
+      raw = value.raw,
+      esc = value.esc,
+    }
+  end)
+end
+
+S.toplevel = P.lazy(function()
+  return P.any(S.placeholder, S.tabstop, S.variable, S.choice)
+end)
+
+S.format = P.any(
+  P.map(P.seq(S.dollar, S.int), function(values)
```

Comment:
Does it make sense to use LPEG or treesitter instead of a new DSL?

We use LPEG in some legacy code:

- https://github.com/neovim/neovim/blob/687a0b3d3e8878b6de058b02d84fbd1f78aa2e89/src/nvim/generators/gen_declarations.lua
- https://github.com/neovim/neovim/blob/687a0b3d3e8878b6de058b02d84fbd1f78aa2e89/src/nvim/generators/c_grammar.lua

Reply:
See my comment above (TL;DR yes (ideally tree-sitter), but we should fix this now before we worry about a better implementation.)

More important is to get an API/format that snippet engines can rely on.

Reply:
(and for the sake of completeness: right now `lpeg` is a build dependency but not a runtime dependency; unless you build from source yourself, it's not bundled.)

Reply:
Another option could be to "vendorize" LPegLJ (or LuLPeg) so there is no need for including more distribution time dependencies.

Reply:
> Another option could be to "vendorize" LPegLJ (or LuLPeg) 

If we're not modifying the source I'm not sure we should do this as it will anger the Debian gods

Reply:
Yes. I thought we should use lpeg for it.
But at this point, I'm not thinking so now. because lpeg is not bundled now and lpeg might be only used as a "snippet" in the functionality provided to the user.